platform/FileHandle.h

```diff
@@ -152,8 +162,106 @@ class FileHandle {
      */
     MBED_DEPRECATED_SINCE("mbed-os-5.4", "Replaced by FileHandle::size")
     virtual off_t flen() { return size(); }
+
+    /** Set blocking or non-blocking mode of the file operation like read/write.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  The default is blocking.
+     *
+     *  @param blocking true for blocking mode, false for non-blocking mode.
+     */
+    virtual int set_blocking(bool blocking) {
+        return -1;
+    }
+
+    /** Check for poll event flags
+     * The input parameter can be used or ignored - the could always return all events,
+     * or could check just the events listed in events.
+     * Call is non-blocking - returns instantaneous state of events.
+     * Whenever an event occurs, the derived class must call _poll_change().
+     * @param events bitmask of poll events we're interested in - POLLIN/POLLOUT etc.
+     * @return bitmask of poll events that have occurred.
+     */
+    virtual short poll(short events) const {
+        // Possible default for real files
+        return MBED_POLLIN | MBED_POLLOUT;
+    }
+    /** Returns true if the FileHandle is writable.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  For example, if the FileHandle is of type Stream, writable() could return
+     *  true when there is ample buffer space available for write() calls.
+     */
+    bool writable() const { return poll(MBED_POLLOUT) & MBED_POLLOUT; }
+
+    /** Returns true if the FileHandle is readable.
+     *  Definition depends upon the subclass implementing FileHandle.
+     *  For example, if the FileHandle is of type Stream, readable() could return
+     *  true when there is something available to read.
+     */
+    bool readable() const { return poll(MBED_POLLIN) & MBED_POLLIN; }
+
+    /** Register a callback on state change of the file operation like read/write
+     *
+     *  The specified callback will be called on state changes such as when
+     *  the file can be written to or read from.
+     *
+     *  The callback may be called in an interrupt context and should not
+     *  perform expensive operations.
+     *
+     *  @param func     Function to call on state change
+     */
+    int attach(Callback<void(short events)> func);
```

Comment:
It could be sigio, yes. At the minute, it is a reliable callback, unlike sigio, but perhaps it shouldn't be documented that way. That would ease socket integration. In that case remove the parameter too.